### PR TITLE
Refactor on `config_manager.py`

### PR DESCRIPTION
**Pull request by chenyushuo**
## Description

Refactor on `config_manager.py`:
1. Add `beginer mode`.
2. Simplify configs.
3. Add more help message on configs.
4. Rename `total_epoch` to `total_epochs`.
5. Add some veRL configs (`norm_adv_by_std_in_grpo`, `use_kl_in_reward`, `horizon` and `target_kl`).
6. `buffer.train_dataset.algorithm_type` will be set to `trainer.algorithm_type` automatically, and `buffer.sft_warmup_dataset.algorithm_type` will be set to `AlgorithmType.SFT` automatically.
7. Support setting `trainer_config` directly without providing `trainer_config_path`.
8. `algorithm_type` in dataset will be set automatically.

## Checklist

Please check the following items before code is ready to be reviewed.

- [x]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: trinity/manager/config_manager.py [hunk 1]
```diff
@@ -1,583 +1,900 @@
...
+
+    def _set_dataset_args(self):
+        if st.session_state["dataset_path"] and "://" not in st.session_state["dataset_path"]:
+            train_split_col, eval_split_col, prompt_key_col, response_key_col = st.columns(4)
+            train_split_col.text_input("Train Split", key="train_split")
+            eval_split_col.text_input("Eval Split", key="eval_split")
+            prompt_key_col.text_input("Prompt Key", key="prompt_key")
+            response_key_col.text_input("Response Key", key="response_key")
+
+    def _set_default_workflow_type(self):  # TODO: add help messages
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: `simple_workflow`: call 'model.chat()' to get responses
`math_workflow`: call 'model.chat()' with a pre-defined system prompt to get responses
Other workflows: conduct multi-turn task for the given dataset

File: trinity/manager/config_manager.py [hunk 2]
```diff
@@ -1,583 +1,900 @@
...
+        st.selectbox(
+            "Default Workflow Type",
+            WORKFLOWS.modules.keys(),
+            key="default_workflow_type",
+        )
+
+    def _set_default_reward_fn_type(self):  # TODO: add help messages
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: `accuracy_reward`: check the accuracy for math problems
`format_reward`: check if the response matches the format (default: `<think>*</think>* <answer>*</answer>`)
`math_reward`: `accuracy_reward` (1 or 0) + `format_reward` (+0.1 or -0.1)

File: trinity/manager/config_manager.py [hunk 3]
```diff
@@ -1,583 +1,900 @@
...
+        )
+
+    def _set_repeat_times(self):
+        st.number_input("Repeat Times", key="repeat_times", min_value=1)
+
+    def _set_sync_method(self):
+        st.selectbox("Sync Method", ["online", "offline"], key="sync_method")
+
+    def _set_sync_iteration_interval(self):
+        st.number_input("Sync Iteration Interval", key="sync_iteration_interval", min_value=1)
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: The iteration interval at which the explorer and trainer synchronize and switch

File: trinity/manager/config_manager.py [hunk 4]
```diff
@@ -1,583 +1,900 @@
...
+    def _set_tensor_parallel_size(self):
+        st.number_input(
+            "Tensor Parallel Size", key="tensor_parallel_size", min_value=1, max_value=8
+        )
+
+    def _set_repeat_times(self):
+        st.number_input("Repeat Times", key="repeat_times", min_value=1)
+
+    def _set_sync_method(self):
+        st.selectbox("Sync Method", ["online", "offline"], key="sync_method")
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: `online`: the explorer and trainer switch at `sync_iter_interval`
`offline`: the trainer saves the model checkpoint, and the explorer loads it at `sync_iter_interval`

File: trinity/manager/config_manager.py [hunk 5]
```diff
@@ -1,583 +1,900 @@
...
+        st.selectbox("Dtype", ["float16", "bfloat16", "float32"], key="dtype")
+
+    def _set_backend(self):
+        st.selectbox("Backend", ["nccl"], key="backend")
+
+    def _set_temperature(self):
+        st.number_input("Temperature", key="temperature", min_value=0.0, max_value=2.0)
+
+    def _set_top_p(self):
+        st.number_input("Top P", key="top_p", min_value=0.0, max_value=1.0)
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: Top P -> Top-p

File: trinity/manager/config_manager.py [hunk 6]
```diff
@@ -1,583 +1,900 @@
...
+
+    def _set_top_k(self):
+        st.number_input(
+            "Top K",
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: Top K -> Top-k

File: trinity/manager/config_manager.py [hunk 7]
```diff
@@ -1,583 +1,900 @@
...
+            min_value=-1,
+            max_value=512,
+            help="Integer that controls the number of top tokens to consider. Set to -1 to consider all tokens.",
+        )
+
+    def _set_seed(self):
+        st.number_input("Seed", key="seed", step=1)
+
+    def _set_logprobs(self):
+        st.number_input("Logprobs", key="logprobs", min_value=0, max_value=20)
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py:477]: does this work?

File: trinity/manager/config_manager.py [hunk 8]
```diff
@@ -1,583 +1,900 @@
...
+        st.number_input("Max Actor Checkpoint to Keep", key="max_actor_ckpt_to_keep", min_value=1)
+
+    def _set_max_critic_ckpt_to_keep(self):
+        st.number_input("Max Critic Checkpoint to Keep", key="max_critic_ckpt_to_keep", min_value=1)
+
+    def _set_gamma(self):
+        st.number_input("Gamma", key="gamma")
+
+    def _set_lam(self):
+        st.number_input("lam", key="lam")
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: lambda? Not sure

File: trinity/manager/config_manager.py (hunk 7 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py:477]: https://github.com/modelscope/Trinity-RFT/blob/main/trinity/common/models/vllm_async_model.py#L57

File: trinity/manager/config_manager.py (hunk 8 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py]: Yes, it is used in veRL.

File: trinity/manager/config_manager.py [hunk 9]
```diff
@@ -1,838 +1,1304 @@
...
-            )
-            st.session_state.engine_num = engine_num
-            tensor_parallel_size = tensor_parallel_size_col.number_input(
-                "Tensor Parallel Size", value=1, min_value=1, max_value=8
+    def _set_sync_method(self):
+        st.selectbox(
+            "Sync Method",
+            ["online", "offline"],
+            key="sync_method",
+            help="""`online`: the explorer and trainer switch at sync_iter_interval.
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py]: "the explorer and trainer synchronize model weights once every sync_iter_interval steps"

File: trinity/manager/config_manager.py (hunk 8 above)
Comment (hiyuchang) [reply trinity/manager/config_manager.py]: Here use lambda is better for understanding.

File: trinity/manager/config_manager.py [hunk 10]
```diff
@@ -1,838 +1,1312 @@
...
+
+`offline`: the trainer saves the model checkpoint, and the explorer loads it at sync_iter_interval.""",
+        )
+
+    def _set_sync_iteration_interval(self):
+        st.number_input(
+            "Sync Iteration Interval",
+            key="sync_iteration_interval",
+            min_value=1,
+            help="""The iteration interval at which the `explorer` and `trainer` synchronize and switch.""",
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py]: Perhaps delete "and switch"? Because explorer and trainer might run simultaneously, with pipeline parallelism.

File: trinity/manager/config_manager.py [hunk 11]
```diff
@@ -1,838 +1,1312 @@
...
-            buffer_advanced_tab = st.expander("Advanced Config")
-            with buffer_advanced_tab:
-                db_url = st.text_input(
-                    "DB URL",
-                    "",
-                    help=r"Default to `sqlite:///{os.path.join(checkpoint_path, '.cache', project, name)}/data.db`",
-                )
-                if not db_url.strip():
-                    db_url = rf"sqlite:///{os.path.join(checkpoint_path, '.cache', project, name)}/data.db"
+    def _set_engine_num(self):
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py:356]: This function seems unused

File: trinity/manager/config_manager.py [hunk 12]
```diff
@@ -1,838 +1,1312 @@
...
+            max_value=max_engine_num,
+            help=f"`engine_num` is used to set the quantity of inference engines, "
+            f"{self._str_for_engine_num_and_tp_size}",
+            on_change=self._set_trainer_gpu_num,
+        )
 
-                max_retry_times_col, max_retry_interval_col = st.columns(2)
-                max_retry_times = max_retry_times_col.number_input(
-                    "Max Retry Times", value=3, min_value=1
+    def _set_tensor_parallel_size(self):
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py:372]: This function seems unused

File: trinity/manager/config_manager.py (hunk 8 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py]: OK

File: trinity/manager/config_manager.py (hunk 9 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py]: OK, fixed.

File: trinity/manager/config_manager.py (hunk 11 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py:356]: It can be called by `_set_configs_with_st_columns`

File: trinity/manager/config_manager.py [hunk 13]
```diff
@@ -1,838 +1,1312 @@
...
+            ["kl", "abs", "mse", "low_var_kl"],
+            key="actor_kl_loss_type",
+        )
+
+    def _set_actor_tau(self):
+        st.number_input(
+            "Tau for OPMD",
+            key="actor_tau",
+            min_value=0.0,
+            max_value=1.0,
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py]: There is no limit for the max value of tau.

File: trinity/manager/config_manager.py [hunk 14]
```diff
@@ -1,838 +1,1312 @@
...
+            format="%.1e",
+        )
+
+    def _set_actor_opmd_baseline(self):
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py:697]: Currently this parameter is ineffective, thus this function might be deleted.

File: trinity/manager/config_manager.py [hunk 15]
```diff
@@ -1,838 +1,1312 @@
...
+        st.selectbox(
+            "OPMD Baseline",
+            ["mean", "logavgexp"],
+            key="actor_opmd_baseline",
+        )
+
+    def _set_actor_use_uid(self):
```
Comment (yanxi-chen) [inline trinity/manager/config_manager.py:704]: Currently this parameter is ineffective, thus this function might be deleted.

File: trinity/manager/config_manager.py (hunk 10 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py]: OK, I rewrite this sentence to """The iteration interval at which the `explorer` and `trainer` synchronize model weight."""

File: trinity/manager/config_manager.py (hunk 13 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py]: ok

File: trinity/manager/config_manager.py [hunk 16]
```diff
@@ -1,838 +1,1326 @@
...
+    def _set_enforce_eager(self):
+        st.checkbox("Enforce Eager", key="enforce_eager")
+
+    def _set_trainer_type(self):
+        st.selectbox("Trainer Type", ["verl"], key="trainer_type")
+
+    def _set_algorithm_type(self):
+        st.selectbox(
+            "Algorithm Type",
+            [AlgorithmType.PPO.value, AlgorithmType.DPO.value, AlgorithmType.OPMD.value],
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py:491]: Comment: For algorithms except DPO and OPMD, please choose PPO and identify `adv_estimater`.

File: trinity/manager/config_manager.py [hunk 17]
```diff
@@ -1,838 +1,1326 @@
...
+        st.number_input("Max Actor Checkpoint to Keep", key="max_actor_ckpt_to_keep", min_value=1)
+
+    def _set_max_critic_ckpt_to_keep(self):
+        st.number_input("Max Critic Checkpoint to Keep", key="max_critic_ckpt_to_keep", min_value=1)
+
+    def _set_gamma(self):
+        st.number_input("Gamma", key="gamma")
+
+    def _set_lam(self):
+        st.number_input("lambda", key="lam")
```
Comment (hiyuchang) [inline trinity/manager/config_manager.py]: lambda -> Lambda
(for consistency)

File: trinity/manager/config_manager.py (hunk 16 above)
Comment (chenyushuo) [reply trinity/manager/config_manager.py:491]: I will handle this in the next PR, and `AlgorithmType.GRPO` will be able to use at that time.

File: trinity/common/config.py [hunk 18]
```diff
@@ -266,21 +266,29 @@ def _check_buffer(self) -> None:
...
             if self.buffer.train_dataset is None:
                 raise ValueError("buffer.train_dataset is required when mode is not 'both'")
-            if self.buffer.train_dataset.algorithm_type != self.trainer.algorithm_type:
-                raise ValueError(
-                    f"buffer.train_dataset.algorithm_type ({self.buffer.train_dataset.algorithm_type}) "
-                    f"is not consistent with trainer.algorithm_type ({self.trainer.algorithm_type})"
-                )
+            self.buffer.train_dataset.algorithm_type = self.trainer.algorithm_type
+        if self.buffer.sft_warmup_dataset is not None:
+            self.buffer.sft_warmup_dataset.algorithm_type = AlgorithmType.SFT
```
Comment (pan-x-c) [inline trinity/common/config.py:271]: Check whether `sft_warmup_dataset.algorithm_type` is other algorithm

Comment (chenyushuo) [reply trinity/common/config.py:271]: In the current design, `sft_warmup_dataset` is only used for SFT, and there are no other values for `algorithm_type `.

**Review by pan-x-c — APPROVED** [review]
